core/src/main/scala/org/apache/spark/status/AppStatusListener.scala

```diff
@@ -747,6 +755,19 @@ private[spark] class AppStatusListener(
         maybeUpdate(esummary, now)
       }
 
+      val speculationStageSummary = stage.speculationStageSummary
+      if (event.taskInfo.speculative) {
+        speculationStageSummary.numActiveTasks -= 1
+        speculationStageSummary.numCompletedTasks += completedDelta
+        speculationStageSummary.numFailedTasks += failedDelta
+        speculationStageSummary.numKilledTasks += killedDelta
+      }
+      if (isLastTask && event.taskInfo.speculative) {
+        update(speculationStageSummary, now)
+      } else if (stage.speculationStageSummary.numTasks > 0) {
+        maybeUpdate(speculationStageSummary, now)
+      }
```

Comment:
```suggestion
      if (speculationStageSummary.numTasks > 0) {
        if (isLastTask) {
          update(speculationStageSummary, now)
        } else {
          maybeUpdate(speculationStageSummary, now)
        }
      }
```

Reply:
I changed `maybeUpdate` to `update` when we encounter a speculative task. The problem with the previous approach was that - if we have a speculative task, then future task end events would all make unnecessary speculation summary writes to DB. What do you think ?